WebXRSession should not hold a unique ptr to WebXRViewerSpace which is ref counted

achan00 opened the pull request:
#### 67e35fd334d59ec2e9d3294b843884dcecdca493
<pre>
WebXRSession should not hold a unique ptr to WebXRViewerSpace which is ref counted
<a href="https://bugs.webkit.org/show_bug.cgi?id=264200">https://bugs.webkit.org/show_bug.cgi?id=264200</a>
<a href="https://rdar.apple.com/115180471">rdar://115180471</a>

Reviewed by Chris Dumez.

WebXRViewerSpace subclasses EventTarget and therefore should be ref-counted.

* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::WebXRSession):
* Source/WebCore/Modules/webxr/WebXRSession.h:
* Source/WebCore/Modules/webxr/WebXRSpace.h:
(WebCore::WebXRViewerSpace::create):

Canonical link: <a href="https://commits.webkit.org/270230@main">https://commits.webkit.org/270230@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/1d5d803a9022e6e38fc4d714514f7cab810e40cd

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/24915 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/3458 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/26168 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/27031 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/22867 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/25183 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/5147 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/897 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/23164 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/25159 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/2488 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/21496 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/27611 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/2185 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/22432 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/28575 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/22718 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/22781 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/26400 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/2140 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/437 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/3419 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/5966 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/2583 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/2480 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

cdumez commented on Source/WebCore/Modules/webxr/WebXRSession.h:
```diff
@@ -99,7 +99,7 @@ class WebXRSession final : public RefCounted<WebXRSession>, public EventTarget,
     const Vector<PlatformXR::Device::ViewData>& views() const { return m_views; }
     const PlatformXR::Device::FrameData& frameData() const { return m_frameData; }
-    const WebXRViewerSpace& viewerReferenceSpace() const { return *m_viewerReferenceSpace; }
+    const WebXRViewerSpace& viewerReferenceSpace() const { return m_viewerReferenceSpace.get(); }
```
I believe `.get()` is unnecessary.

cdumez commented on Source/WebCore/Modules/webxr/WebXRSpace.h:
```diff
@@ -73,18 +73,21 @@ class WebXRSpace : public EventTarget, public ContextDestructionObserver {
     WebXRViewerSpace(Document&, WebXRSession&);
     virtual ~WebXRViewerSpace();
 
+    using RefCounted<WebXRViewerSpace>::ref;
```
I think you can omit `<WebXRViewerSpace>`

cdumez commented on Source/WebCore/Modules/webxr/WebXRSpace.h:
```diff
@@ -73,18 +73,21 @@ class WebXRSpace : public EventTarget, public ContextDestructionObserver {
     virtual ~WebXRViewerSpace();
 
+    using RefCounted<WebXRViewerSpace>::ref;
+    using RefCounted<WebXRViewerSpace>::deref;
```
ditto.

cdumez commented on Source/WebCore/Modules/webxr/WebXRSpace.h:
```diff
@@ -73,18 +73,21 @@ class WebXRSpace : public EventTarget, public ContextDestructionObserver {
+class WebXRViewerSpace : public RefCounted<WebXRViewerSpace>, public WebXRSpace {
     WTF_MAKE_ISO_ALLOCATED(WebXRViewerSpace);
 public:
     WebXRViewerSpace(Document&, WebXRSession&);
```
Refcounted classes should never have a public constructor. That too error prone. Please make the constructor private and add a public factory function instead:
```
static Ref< WebXRViewerSpace> create(Document& document, WebXRSession& session)
{
    return adoptRef(*new WebXRViewerSpace(document, session));
}
```

cdumez commented on Source/WebCore/Modules/webxr/WebXRSession.cpp:
```diff
@@ -64,7 +64,7 @@ WebXRSession::WebXRSession(Document& document, WebXRSystem& system, XRSessionMod
     , m_requestedFeatures(WTFMove(requestedFeatures))
     , m_activeRenderState(WebXRRenderState::create(mode))
-    , m_viewerReferenceSpace(makeUnique<WebXRViewerSpace>(document, *this))
+    , m_viewerReferenceSpace(adoptRef(*new WebXRViewerSpace(document, *this)))
```
Use WebXRViewerSpace::create()

cdumez reviewed (approved):
LGTM

achan00 committed:
WebXRSession should not hold a unique ptr to WebXRViewerSpace which is ref counted
https://bugs.webkit.org/show_bug.cgi?id=264200
rdar://115180471

Reviewed by Chris Dumez.

WebXRViewerSpace subclasses EventTarget and therefore should be ref-counted.

* Source/WebCore/Modules/webxr/WebXRSession.cpp:
(WebCore::WebXRSession::WebXRSession):
* Source/WebCore/Modules/webxr/WebXRSession.h:
* Source/WebCore/Modules/webxr/WebXRSpace.h:
(WebCore::WebXRViewerSpace::create):

Canonical link: https://commits.webkit.org/270230@main
